tests/rptest/services/action_injector.py

```diff
@@ -0,0 +1,279 @@
+            self.last_affected_node = node
+
+            # Update started_nodes so storage validations are run
+            # on the correct set of nodes later.
+            self.redpanda.remove_from_started_nodes(node)
+            return node
+        else:
+            self.redpanda.logger.warn(f'no usable node')
+
+    def do_reverse_action(self):
+        self._start_rp(node=self.last_affected_node)
+        self.affected_nodes.remove(self.last_affected_node)
+        self.redpanda.add_to_started_nodes(self.last_affected_node)
+
+        last_affected_node, self.last_affected_node = self.last_affected_node, None
+        return last_affected_node
+
+    def do_restore_nodes(self, nodes_to_restore: Set[ClusterNode]):
+        """
+        Attempt to restore the redpanda process on all nodes where it was stopped.
+        """
+        for node in nodes_to_restore:
+            self._start_rp(node)
+
+    def _start_rp(self, node):
+        self.failure_injector._start(node)
+        wait_until(
+            lambda: self.redpanda.redpanda_pid(node),
+            timeout_sec=self.PROCESS_START_WAIT_SEC,
+            backoff_sec=self.PROCESS_START_WAIT_BACKOFF,
+            err_msg=
+            f'Failed to start redpanda process on {node.account.hostname}')
+
+
+class ActionInjectorThread(Thread):
+    def __init__(
+        self,
+        config: ActionConfig,
+        redpanda: RedpandaService,
+        disruptive_action: DisruptiveAction,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.disruptive_action = disruptive_action
+        self.redpanda = redpanda
+        self.config = config
+        self._stop_requested = Event()
+        self.action_log = []
+
+    def run(self):
+        admin = Admin(self.redpanda)
+
+        def all_nodes_started():
+            statuses = [
+                admin.ready(node).get("status") for node in self.redpanda.nodes
+            ]
+            return all(status == 'ready' for status in statuses)
+
+        wait_until(all_nodes_started,
+                   timeout_sec=self.config.cluster_start_lead_time_sec,
+                   backoff_sec=2,
+                   err_msg=f'Cluster not ready to begin actions')
+
+        self.redpanda.logger.info('cluster is ready, starting action loop')
+
+        while not self._stop_requested.is_set():
+            if result := self.disruptive_action.action():
+                self.action_log.append(
+                    ActionLogEntry(result, is_reverse_action=False))
+
+            time.sleep(self.config.time_between_actions())
+
+            if result := self.disruptive_action.reverse():
+                self.action_log.append(
+                    ActionLogEntry(result, is_reverse_action=True))
+
+        if self.config.restore_state_on_exit:
+            self.redpanda.logger.info('attempting to restore system state')
+            self.disruptive_action.restore_state_on_exit(self.action_log)
```

Comment:
It seems that this code doesn't do anything: we already invoke `.reverse` for each `.action` so by the time we reached this method all the actions are already reversed.

As a result we don't need:
 - ActionConfig::restore_state_on_exit (config)
 - DisruptiveAction::restore_state_on_exit (method)
 - `action_log` (tracking history of a DisruptiveAction's actions inside ActionInjectorThread)
 - return nodes from `.reverse` and `.action`
 - introduce ActionLogEntry

It's better avoid code which we don't need

Reply:
Fair point. When I added the restore action I was thinking of operations which are not reversible but still need some kind of restore on exit. But now that you pointed it out I realised that there are no such operations that will require such a restore at this time that I can think of.

If there are such operations in future, I will revisit, but I will remove this code for now. 

Action log might still be required though because it allows assertions by unit test code easily to make sure that the code is called. 

I will check if some intercept can be used to assert that the action injector has worked, otherwise I might need to keep the log for assertions (or some simpler mechanism than a log).

Reply:
removed the restore related code in latest commit. for making assertions easy I have changed it to two flags and an assert helper on the action context, and the log is removed